docs/blocks/states.md

```diff
@@ -1,96 +1,127 @@
+    }
+}
 ```
 
-If you wish to change what `BlockState` is used when placing your block, you can overwrite `Block#getStateForPlacement(BlockPlaceContext)`. This can be used to, for example, set the direction of your block depending on where the player is standing when they place it.
+Using Blockstates
+-----------------
+
+To go from `Block` to `BlockState`, call `Block#defaultBlockState()`.
+
+You can get the value of a property by calling `BlockState#getValue(Property<?>)`, passing it the property you want to get the value of. Reusing our lever example, this would look something like `Direction direction = leverBlockState.getValue(LeverBlock.FACING);`
+
+If you want to get a `BlockState` with a different set of values, simply call `BlockState#setValue(Property<T>, T)` on an existing block state with the property and its value. With our lever, this goes something like `leverBlockState = leverBlockState.setValue(LeverBlock.FACING, Direction.SOUTH);`
+
+:::note
+`BlockState`s are immutable. This means that when you call `#setValue(Property<T>, T)`, you are not actually modifying the blockstate. Instead, a lookup is performed internally, and you are given the blockstate object you requested, which is the one and only object that exists with these exact property values. This also means that just calling `state#setValue` without saving it into a variable (for example back into `state`) does nothing.
+:::
+
+To get a `BlockState` from the level, use `Level#getBlockState(BlockPos)`.
 
-Because `BlockState`s are immutable, and all combinations of their properties are generated on startup of the game, calling `BlockState#setValue(Property<T>, T)` will simply go to the `Block`'s `StateHolder` and request the `BlockState` with the set of values you want.
+### `Level#setBlock`
 
-Because all possible `BlockState`s are generated at startup, you are free and encouraged to use the reference equality operator (`==`) to check if two `BlockState`s are equal.
+To set a `BlockState` in the level, use `Level#setBlock(BlockPos, BlockState, int)`.
 
-Using `BlockState`'s
----------------------
+The `int` parameter deserves some extra explanation, as its meaning is not immediately obvious. It denotes what is known as update flags.
 
-You can get the value of a property by calling `BlockState#getValue(Property<?>)`, passing it the property you want to get the value of.
-If you want to get a `BlockState` with a different set of values, simply call `BlockState#setValue(Property<T>, T)` with the property and its value.
+To help setting the update flags correctly, there are a number of `int` constants in `Block`, prefixed with `UPDATE_`. Examples include `Block.UPDATE_NEIGHBORS` or `Block.UPDATE_CLIENTS`. These constants can be added together (for example `Block.UPDATE_NEIGHBORS + Block.UPDATE_CLIENTS`) if you wish to combine them. If you're unsure which one to use, use `Block.UPDATE_ALL`.
```

Comment:
Probably best to explain what all does since it doesn't really update 'all'.

Reply:
Being completely honest, I don't exactly know what each of these do myself. Help would be appreciated for this one.

Reply:
They are sort of self explanatory, but all of the logic is in `Level#markAndNotifyBlock`, which I believe is Forge added to split the logic.

`UPDATE_NEIGHBORS` basically update the surrounding six directions by calling `#neighborChanged`. From there it additionally calls `#onNeighborChange` and `#isRedstoneConductor`, if `#hasAnalogOutputSignal` return true. If the redstone conductor check succeeds, then `#getWeakChanges` is called to determine whether the block two away from the updating one should call `#neighborChanged`.

As for `UPDATE_CLIENTS`, it sends a packet on the next tick containing the blocks to change on the client.

As an additional check, if `UPDATE_KNOWN_SHAPE` isn't set and the depth is still greater than 0, it essentially recursively checks the surrounding blocks and performs updates until everything has updated. For reference, the flags passed in to the update shapes method zero out `UPDATE_NEIGHBORS`, `UPDATE_CLIENTS`, `UPDATE_SUPPRESS_DROPS`, and `UPDATE_MOVE_BY_PISTON`. While this means that the blocks themselves won't be sent through a packet on the client, they are rerendered immediately.

`UPDATE_ALL` combines everything above.

Reply:
Added (hopefully correctly xd)